[doc] fixing dead links + tool to check for dead links automatically

kris-scheibe opened the pull request:
I found a few dead links in the documentation, so I fixed them 😄

The tool I used is now part of the regular `pmd-doc` build, i.e. the build fails if any dead links are detected. The detection currently fails on links to the CLI reference as the anchors to the parameters are defined as plain html and not md captions, so links to the CLI reference aren't checked right now.

Here is the output of the initial run (this includes a some already fixed issues (#1542, dfd5b7b8865d984feefaed331a31b1c9fae5fb2e) as my fork was a few commits behind):
```
Found dead link(s):
/home/kris/Checkout/pmd/docs/pages/pmd/rules/apex.md
     117: [ApexDoc](pmd_rules_apex_documentation.html#apexdoc)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/extending/writing_pmd_rules.md
     251: [here](pmd-core/xref/net/sourceforge/pmd/lang/ast/xpath/AttributeAxisIterator.html)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/extending/rule_guidelines.md
      53: [Test Framework](pmd_devdocs_testing.html)
      62: [Using the test framework externally](pmd_devdocs_testing.html#using-the-test-framework-externally)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/extending/writing_xpath_rules.md
      19: [here](howtowritearule.html)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/suppressing_warnings.md
      35: [How to write a rule](pmd_devdocs_writing_pmd_rules.html)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/best_practices.md
      24: [via a custom ruleset](pmd_userdocs_understanding_rulesets.html)
      28: [suppress](pmd_userdocs_suppressing.html)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/cpd.md
      20: [here](/pmd_devdocs_adding_new_cpd_language.html)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/cli_reference.md
     169: [ecmascript](pmd_rules_javascript.html)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/tools/maven.md
     112: [here](pmd_userdocs_understanding_rulesets.html)
     239: [here](/pmd_userdocs_understanding_rulesets.html)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/installation.md
     108: [Supported Languages](pmd_userdocs_cpd#supported-languages)
/home/kris/Checkout/pmd/docs/pages/pmd/projectdocs/credits.md
     431: [howto for adding new languages](../customizing/new-language.html)
/home/kris/Checkout/pmd/docs/pages/release_notes_old.md
     672: [`AccessorClassGeneration`](pmd_rules_java_bestpracices.html#accessorclassgeneration)
     673: [`AccessorMethodGeneration`](pmd_rules_java_bestpracices.html#accessormethodgeneration)
    1146: [Incremental Analysis](pmd_userdocs_getting_started.html)
    1770: [Working with properties](pmd_devdocs_working_with_properties.html)
[WARNING] 
java.lang.AssertionError: dead links detected
    at net.sourceforge.pmd.docs.DeadLinksChecker.checkDeadLinks (DeadLinksChecker.java:100)
    ...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: No java rules are changed!
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

oowekyala commented on pmd-doc/src/main/java/net/sourceforge/pmd/docs/GenerateRuleDocsCmd.java:
```diff
@@ -41,6 +41,9 @@ public static void main(String[] args) throws RuleSetNotFoundException {
         System.out.println("Generated docs in " + (System.currentTimeMillis() - start) + " ms");
+
+        DeadLinksChecker deadLinksChecker = new DeadLinksChecker();
+        deadLinksChecker.checkDeadLinks(output.resolve("docs/pages"));
```
Also I wonder whether we could check http links for 404s and 410s... E.g. with https://stackoverflow.com/questions/1378199/how-to-check-if-a-url-exists-or-returns-404-with-java

kris-scheibe commented:
I initially wrote this as a test.
The thing is, that the tests run before the docs are being generated and I wanted the check to run on the newly generated documents.

Also the generator is run by the maven-exec plugin and derives the base directory from the `${project.basedir}` Maven property -- I think the dead link checker should be run the same way, but I can move it to the verify phase easily.

Testing the external links is fairly easy, but I don't think it's feasible to check them for any regular build.
* Testing a lot of links takes a _long_ time... my initial run took about 20min. (I've optimized that quite a bit, but it still takes 4-5min on my machine building the pmd-doc module when checking external links.)
* Sometimes checking fails even if the link is generally valid, e.g. on one of my first runs all requests to `javadoc.io` failed because they hadn't had the PMD javadocs in their cache, and once I might've triggered the DDoS protection of GitHub.
* And most importantly: Due to the two points above enabling the the dead link checker for external links by default would randomly break the PMD build, even if the documentation didn't change.

So I implemented checking external links with a few optimizations:
* Checking external links depends on the `pmd.doc.checkExternalLinks` property being set.
* Links to `https://github.com/pmd/pmd/blob/master/` will be checked against the local checked out file instead of the current master branch on GitHub.
* Links to PMD PR's and Issues will not be checked at all. They constitute the bulk of links and are pretty much impossible to disappear as long as PMD is on GitHub.
* Every link is only checked once (especially the GitHub pages of regular committers appear quite often).

I fixed a few of the more important links but I think others should be discussed:
* the OmniCore IDE and InfoZip projects are either dead or haven't had any updates in years -> remove?
* personal home pages / GitHub pages (in credits and release-notes) -> remove link, keep plain-text URL?
* known broken links -> remove link, keep plain-text URL?
```
~/Checkout/pmd$ ./mvnw clean verify -pl pmd-doc -Dpmd.doc.checkExternalLinks=true

...

Scanned 98 files for dead links.
  Found 1721 external links, 538 of those where checked.
Found dead link(s):
/home/kris/Checkout/pmd/docs/pages/pmd/about/help.md
      19: [archive2](http://blog.gmane.org/gmane.comp.java.audit.pmd.devel)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/tools/maven.md
     201: [here](http://jira.codehaus.org/browse/MPMD)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/tools/tools.md
     192: [check it out](http://www.omnicore.com/)
/home/kris/Checkout/pmd/docs/pages/pmd/userdocs/installation.md
      18: [InfoZip](http://www.info-zip.org/pub/infozip/)
/home/kris/Checkout/pmd/docs/pages/pmd/projectdocs/credits.md
     190: [Wim Deblauwe](http://www.jroller.com/page/Fester)
     224: [Lori Olson](http://blogs.teamb.com/LoriOlson/)
     278: [Todd Wright](http://freshmeat.net/~toddwright/)
     323: [Vladimir Bossicard](http://www.bossicard.com/)
/home/kris/Checkout/pmd/docs/pages/pmd/projectdocs/trivia/news.md
      16: [Free Software Magazine](http://www.freesoftwaremagazine.com/columns/destroy_annoying_bugs_part_1)
      23: [Software Testing and Performance Magazine [link broken]](http://www.stpmag.com/issues/stp-2006-10.pdf)
      79: [JJGuidelines - Java and J2EE Conventions, Guidelines and Best Practices [link broken]](https://jjguidelines.dev.java.net/index.html)
/home/kris/Checkout/pmd/docs/pages/pmd/devdocs/major_contributions/adding_new_metrics_framework.md
     103: [Example](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/metrics/AbstractJavaClassMetric.java#L50)
/home/kris/Checkout/pmd/docs/pages/release_notes_old.md
     463: [`ApexDoc`](https://pmd.github.io/pmd-6.8.0/pmd_rules_apex_documentation.html#apexdoc)
    1337: [core] Support shortnames when using filelist - [John Zhang](https://github.com/johnjiabinzhang)
    1419: [apex] AvoidDirectAccessTriggerMap incorrectly detects array access in classes - [Robert S?semann](https://github.com/up2go-rsoesemann)
    1420: [apex] Method chaining breaks ApexCRUDViolation - [Robert S?semann](https://github.com/up2go-rsoesemann)
    1943: [apex] Make Rule suppression work - [Robert S?semann](https://github.com/up2go-rsoesemann)
    3275: [apex #43](https://github.com/Up2Go/pmd/pull/43)
[WARNING] 
java.lang.AssertionError: dead links detected
    ...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 04:08 min
```

oowekyala commented:
This looks good to me. Happy new year btw!

I pushed a patch that makes the evaluation of the http requests parallel. That makes the program run in about 15 secs on my machine ^^ Also I cleaned up the log output, but whatever

> the OmniCore IDE and InfoZip projects are either dead or haven't had any updates in years -> remove?
> personal home pages / GitHub pages (in credits and release-notes) -> remove link, keep plain-text URL?

I agree with those two

> known broken links -> remove link, keep plain-text URL?

* Links in trivia/news.md can be removed. They were probably news a decade ago lol
* Links in release_notes_old.md can be ignored I think... They're kind of expected to go dead at some point

I'll fix those and update the documentation to mention the task and the maven property... then this should be good to go

kris-scheibe commented:
Happy new year to you too! @oowekyala 

I'm not sure if my internet connection is better or worse than yours, but running the dead links checker now results in a *lot* of connection errors, and a few [HTTP 429](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429)...

```
...
[DEBUG] response: HTTP 200 on https://github.com/adangel/pmd/pull/33
[DEBUG] response: HTTP 429 on https://github.com/oowekyala
[DEBUG] response: java.net.SocketTimeoutException on https://github.com/adangel/pmd/pull/27 : connect timed out
[DEBUG] response: java.net.SocketTimeoutException on https://github.com/adangel/pmd/pull/32 : connect timed out
[DEBUG] response: java.net.SocketTimeoutException on https://github.com/adangel/pmd/pull/25 : connect timed out
[DEBUG] response: HTTP 429 on https://github.com/oowekyala
[DEBUG] response: java.net.SocketTimeoutException on https://github.com/adangel/pmd/pull/25 : connect timed out
[DEBUG] response: java.net.SocketTimeoutException on https://github.com/adangel/pmd/pull/23 : connect timed out
...
```
Also you missed a few System.out.println's starting [here](https://github.com/kris-scheibe/pmd/blob/f922bcac0e682a87e54dfdd733c8a64b7ea3718b/pmd-doc/src/main/java/net/sourceforge/pmd/docs/DeadLinksChecker.java#L178). ;)

oowekyala commented:
@kris-scheibe Good catch, there appeared to be a problem with the response cache. It's fixed in 2973453946c26faadb82ac8f76d969b61bc68591

jsotuyod commented on pmd-doc/src/main/java/net/sourceforge/pmd/docs/DeadLinksChecker.java:
```diff
@@ -0,0 +1,327 @@
+
+        try {
+            final HttpURLConnection httpURLConnection = (HttpURLConnection) new URL(url).openConnection();
+            httpURLConnection.setRequestMethod("GET");
```
since we only care about the status code, we could simply use `HEAD` instead of `GET`

jsotuyod commented on pmd-doc/src/main/java/net/sourceforge/pmd/docs/DeadLinksChecker.java:
```diff
@@ -0,0 +1,327 @@
+            Files.walk(pagesDirectory)
+                 .filter(Files::isRegularFile)
+                 .filter(path -> path.toString().endsWith(".md"))
+                 .forEach(mdFiles::add);
```
`collect(Collectors.toList())`
